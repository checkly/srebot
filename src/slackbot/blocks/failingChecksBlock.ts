import { checkly } from "../../checkly/client";

export type Check = {
  checkId: string;
  checkState: "FAILED" | "DEGRADED" | "PASSED";
  name: string;
  failures: {
    total: number;
    timeframe: string;
  };
  group: string | null;
  lastFailure: {
    checkResultId: string;
    timestamp: Date;
  } | null;
};

export function renderFailingChecksBlock(checks: Check[]) {
  const statusIcon = (checkState: Check["checkState"]) => {
    switch (checkState) {
      case "FAILED":
        return "❌";
      case "DEGRADED":
        return "⚠️";
      case "PASSED":
        return "✅";
    }
  };

  const checkUrl = (checkId: string) => checkly.getCheckAppUrl(checkId);
  const checkResultUrl = (checkId: string, checkResultId: string) =>
    checkly.getCheckResultAppUrl(checkId, checkResultId);
  return {
    blocks: [
      {
        type: "header",
        text: {
          type: "plain_text",
          text: "Failing Checks",
          emoji: true,
        },
      },
      {
        type: "divider",
      },
      {
        type: "section",
        fields: [
          {
            type: "mrkdwn",
            text: "*Check Name*",
          },
          {
            type: "mrkdwn",
            text: "*Failure rate*",
          },
          {
            type: "mrkdwn",
            text: "*Group*",
          },
          {
            type: "mrkdwn",
            text: "*Last failed*",
          },
        ],
      },
      ...checks.map((check) => ({
        type: "section",
        fields: [
          {
            type: "mrkdwn",
            text: `<${checkUrl(check.checkId)}|${statusIcon(check.checkState)} ${check.name}>`,
          },
          {
            type: "mrkdwn",
            text: `${check.failures.total} failure${
              check.failures.total === 1 ? "" : "s"
            } (last ${check.failures.timeframe})`,
          },
          {
            type: "mrkdwn",
            text: check.group || " ",
          },
          {
            type: "mrkdwn",
            text: check.lastFailure
              ? `<!date^${Math.floor(new Date(check.lastFailure.timestamp).getTime() / 1000)}^{ago}|${check.lastFailure.timestamp}> <${checkResultUrl(check.checkId, check.lastFailure.checkResultId)}|Link>`
              : "N/A",
          },
        ],
      })),
    ],
  };
}
