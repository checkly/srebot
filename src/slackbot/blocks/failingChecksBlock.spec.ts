import { Check, renderFailingChecksBlock } from "./failingChecksBlock";

describe("failingChecksBlock", () => {
  const now = new Date();

  it("renders checks with mixed states", () => {
    const checks = [
      {
        checkId: "123",
        checkState: "FAILED",
        name: "Check 1",
        failures: { total: 1, timeframe: "24h" },
        lastFailure: {
          checkResultId: "123",
          timestamp: now,
        },
      },
      {
        checkId: "124",
        checkState: "PASSED",
        name: "Check 2",
        failures: { total: 0, timeframe: "24h" },
        group: "Group 2",
      },
      {
        checkId: "125",
        checkState: "DEGRADED",
        name: "Check 3",
        failures: { total: 2, timeframe: "24h" },
        group: "Group 3",
        lastFailure: {
          checkResultId: "125",
          timestamp: now,
        },
      },
      {
        checkId: "126",
        checkState: "PASSED",
        name: "Check 4",
        failures: { total: 0, timeframe: "24h" },
        group: "Group 4",
      },
    ] as Check[];

    const blocks = renderFailingChecksBlock(checks);
    expect(blocks).toMatchSnapshot();
  });

  it("renders all passing checks", () => {
    const checks = [
      {
        checkId: "124",
        checkState: "PASSED",
        name: "Check 2",
        failures: { total: 0, timeframe: "24h" },
        group: "Group 2",
      },
      {
        checkId: "126",
        checkState: "PASSED",
        name: "Check 4",
        failures: { total: 0, timeframe: "24h" },
        group: "Group 4",
      },
    ] as Check[];

    const blocks = renderFailingChecksBlock(checks);
    expect(blocks).toMatchSnapshot();
  });

  it("renders all failing checks", () => {
    const checks = [
      {
        checkId: "123",
        checkState: "FAILED",
        name: "Check 1",
        failures: { total: 1, timeframe: "24h" },
        lastFailure: {
          checkResultId: "123",
          timestamp: now,
        },
      },
      {
        checkId: "125",
        checkState: "DEGRADED",
        name: "Check 3",
        failures: { total: 2, timeframe: "24h" },
        group: "Group 3",
        lastFailure: {
          checkResultId: "125",
          timestamp: now,
        },
      },
    ] as Check[];

    const blocks = renderFailingChecksBlock(checks);
    expect(blocks).toMatchSnapshot();
  });
});
