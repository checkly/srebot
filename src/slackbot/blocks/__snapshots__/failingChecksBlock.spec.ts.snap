// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`failingChecksBlock renders all failing checks 1`] = `
{
  "blocks": [
    {
      "text": {
        "emoji": true,
        "text": "Failing Checks",
        "type": "plain_text",
      },
      "type": "header",
    },
    {
      "type": "divider",
    },
    {
      "fields": [
        {
          "text": "*Check Name*",
          "type": "mrkdwn",
        },
        {
          "text": "*Failure rate*",
          "type": "mrkdwn",
        },
        {
          "text": "*Group*",
          "type": "mrkdwn",
        },
        {
          "text": "*Last failed*",
          "type": "mrkdwn",
        },
      ],
      "type": "section",
    },
    {
      "fields": [
        {
          "text": "<https://app.checklyhq.com/checks/123|❌ Check 1>",
          "type": "mrkdwn",
        },
        {
          "text": "1 failure (last 24h)",
          "type": "mrkdwn",
        },
        {
          "text": " ",
          "type": "mrkdwn",
        },
        {
          "text": "<!date^1742374341^{ago}|Wed Mar 19 2025 09:52:21 GMT+0100 (Central European Standard Time)> <https://app.checklyhq.com/checks/123/check-sessions/results/123|Link>",
          "type": "mrkdwn",
        },
      ],
      "type": "section",
    },
    {
      "fields": [
        {
          "text": "<https://app.checklyhq.com/checks/125|⚠️ Check 3>",
          "type": "mrkdwn",
        },
        {
          "text": "2 failures (last 24h)",
          "type": "mrkdwn",
        },
        {
          "text": "Group 3",
          "type": "mrkdwn",
        },
        {
          "text": "<!date^1742374341^{ago}|Wed Mar 19 2025 09:52:21 GMT+0100 (Central European Standard Time)> <https://app.checklyhq.com/checks/125/check-sessions/results/125|Link>",
          "type": "mrkdwn",
        },
      ],
      "type": "section",
    },
  ],
}
`;

exports[`failingChecksBlock renders all passing checks 1`] = `
{
  "blocks": [
    {
      "text": {
        "emoji": true,
        "text": "Failing Checks",
        "type": "plain_text",
      },
      "type": "header",
    },
    {
      "type": "divider",
    },
    {
      "fields": [
        {
          "text": "*Check Name*",
          "type": "mrkdwn",
        },
        {
          "text": "*Failure rate*",
          "type": "mrkdwn",
        },
        {
          "text": "*Group*",
          "type": "mrkdwn",
        },
        {
          "text": "*Last failed*",
          "type": "mrkdwn",
        },
      ],
      "type": "section",
    },
    {
      "fields": [
        {
          "text": "<https://app.checklyhq.com/checks/124|✅ Check 2>",
          "type": "mrkdwn",
        },
        {
          "text": "0 failures (last 24h)",
          "type": "mrkdwn",
        },
        {
          "text": "Group 2",
          "type": "mrkdwn",
        },
        {
          "text": "N/A",
          "type": "mrkdwn",
        },
      ],
      "type": "section",
    },
    {
      "fields": [
        {
          "text": "<https://app.checklyhq.com/checks/126|✅ Check 4>",
          "type": "mrkdwn",
        },
        {
          "text": "0 failures (last 24h)",
          "type": "mrkdwn",
        },
        {
          "text": "Group 4",
          "type": "mrkdwn",
        },
        {
          "text": "N/A",
          "type": "mrkdwn",
        },
      ],
      "type": "section",
    },
  ],
}
`;

exports[`failingChecksBlock renders checks with mixed states 1`] = `
{
  "blocks": [
    {
      "text": {
        "emoji": true,
        "text": "Failing Checks",
        "type": "plain_text",
      },
      "type": "header",
    },
    {
      "type": "divider",
    },
    {
      "fields": [
        {
          "text": "*Check Name*",
          "type": "mrkdwn",
        },
        {
          "text": "*Failure rate*",
          "type": "mrkdwn",
        },
        {
          "text": "*Group*",
          "type": "mrkdwn",
        },
        {
          "text": "*Last failed*",
          "type": "mrkdwn",
        },
      ],
      "type": "section",
    },
    {
      "fields": [
        {
          "text": "<https://app.checklyhq.com/checks/123|❌ Check 1>",
          "type": "mrkdwn",
        },
        {
          "text": "1 failure (last 24h)",
          "type": "mrkdwn",
        },
        {
          "text": " ",
          "type": "mrkdwn",
        },
        {
          "text": "<!date^1742374341^{ago}|Wed Mar 19 2025 09:52:21 GMT+0100 (Central European Standard Time)> <https://app.checklyhq.com/checks/123/check-sessions/results/123|Link>",
          "type": "mrkdwn",
        },
      ],
      "type": "section",
    },
    {
      "fields": [
        {
          "text": "<https://app.checklyhq.com/checks/124|✅ Check 2>",
          "type": "mrkdwn",
        },
        {
          "text": "0 failures (last 24h)",
          "type": "mrkdwn",
        },
        {
          "text": "Group 2",
          "type": "mrkdwn",
        },
        {
          "text": "N/A",
          "type": "mrkdwn",
        },
      ],
      "type": "section",
    },
    {
      "fields": [
        {
          "text": "<https://app.checklyhq.com/checks/125|⚠️ Check 3>",
          "type": "mrkdwn",
        },
        {
          "text": "2 failures (last 24h)",
          "type": "mrkdwn",
        },
        {
          "text": "Group 3",
          "type": "mrkdwn",
        },
        {
          "text": "<!date^1742374341^{ago}|Wed Mar 19 2025 09:52:21 GMT+0100 (Central European Standard Time)> <https://app.checklyhq.com/checks/125/check-sessions/results/125|Link>",
          "type": "mrkdwn",
        },
      ],
      "type": "section",
    },
    {
      "fields": [
        {
          "text": "<https://app.checklyhq.com/checks/126|✅ Check 4>",
          "type": "mrkdwn",
        },
        {
          "text": "0 failures (last 24h)",
          "type": "mrkdwn",
        },
        {
          "text": "Group 4",
          "type": "mrkdwn",
        },
        {
          "text": "N/A",
          "type": "mrkdwn",
        },
      ],
      "type": "section",
    },
  ],
}
`;
